src/types.rs

```diff
@@ -558,6 +570,58 @@ impl<T: ToRedisArgs> ToRedisArgs for Option<T> {
     }
 }
 
+impl<T: ToRedisArgs + Hash + Eq> ToRedisArgs for HashSet<T> {
+    fn to_redis_args(&self) -> Vec<Vec<u8>> {
+        let mut rv : Vec<&T> = Vec::with_capacity(self.len());
+
+        rv.extend(self.iter());
+        ToRedisArgs::make_arg_vec_ref(rv.as_slice())
+    }
+
+    fn is_single_arg(&self) -> bool {
+        self.len() <= 1
+    }
+}
+
+impl<T: ToRedisArgs + Hash + Eq + Ord> ToRedisArgs for BTreeSet<T> {
+    fn to_redis_args(&self) -> Vec<Vec<u8>> {
+        let mut rv : Vec<&T> = Vec::with_capacity(self.len());
+
+        rv.extend(self.iter());
+        ToRedisArgs::make_arg_vec_ref(rv.as_slice())
+    }
+
+    fn is_single_arg(&self) -> bool {
+        self.len() <= 1
+    }
+}
+
+/// this flattens BTreeMap into something that goes well with HMSET
+impl<T: ToRedisArgs + Hash + Eq + Ord,
+     V: ToRedisArgs> ToRedisArgs for BTreeMap<T,V> {
+    fn to_redis_args(&self) -> Vec<Vec<u8>> {
+        let mut rv = Vec::with_capacity(self.len());
+
+        rv.extend(self.keys()
+            .zip(self.values())
+            .into_iter()
+            .flat_map(|e| {
```

Comment:
Use value destructuring here: `|(key,value)|`. Makes the code below far easier to read.